File: app/src/main/kotlin/tech/mmarca/openvitals/features/dashboard/components/MetricStatCard.kt
package tech.mmarca.openvitals.features.dashboard.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import tech.mmarca.openvitals.core.presentation.DisplayValue
import tech.mmarca.openvitals.features.dashboard.dashboardDisplayValue
import tech.mmarca.openvitals.ui.components.AutoResizeText
import tech.mmarca.openvitals.ui.components.OpenVitalsCard
import tech.mmarca.openvitals.ui.theme.accentSurfaceContainerColor

private val MetricStatIconSize = 28.dp
private val MetricStatIconGlyphSize = 16.dp
private val MetricStatProgressHeight = 3.dp

@Composable
internal fun MetricStatCard(
    title: String,
    value: DisplayValue,
    icon: ImageVector,
    accentColor: Color,
    modifier: Modifier = Modifier,
    message: String? = null,
    subtitle: String? = null,
    subtitleColor: Color = MaterialTheme.colorScheme.onSurfaceVariant,
    showTitle: Boolean = true,
    progressFraction: Float? = null,
    onClick: (() -> Unit)? = null,
) {
    val iconContainerColor = accentSurfaceContainerColor(
        accentColor = accentColor,
        amoledAlpha = 0.16f,
        fallback = MaterialTheme.colorScheme.surfaceContainerHighest,
    )
    val progressFillColor = accentColor.copy(alpha = 0.55f)

    OpenVitalsCard(
        modifier = modifier.fillMaxWidth(),
        onClick = onClick,
        containerColor = MaterialTheme.colorScheme.surfaceContainer,
    ) {
        Box(modifier = Modifier.fillMaxSize()) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 12.dp, vertical = 10.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Box(
                    modifier = Modifier
                        .size(MetricStatIconSize)
                        .background(color = iconContainerColor, shape = CircleShape),
                    contentAlignment = Alignment.Center,
                ) {
                    Icon(
                        imageVector = icon,
                        contentDescription = null,
                        tint = accentColor,
                        modifier = Modifier.size(MetricStatIconGlyphSize),
                    )
                }
                Spacer(Modifier.width(10.dp))
                Column(
                    modifier = Modifier.weight(1f),
                    verticalArrangement = Arrangement.Center,
                ) {
                    if (showTitle) {
                        AutoResizeText(
                            text = title,
                            style = MaterialTheme.typography.labelMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                            maxLines = 1,
                        )
                    }
                    AutoResizeText(
                        text = message ?: dashboardDisplayValue(value),
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.SemiBold,
                        color = if (message == null) {
                            MaterialTheme.colorScheme.onSurface
                        } else {
                            MaterialTheme.colorScheme.onSurfaceVariant
                        },
                        maxLines = 1,
                    )
                    if (message == null && subtitle != null) {
                        AutoResizeText(
                            text = subtitle,
                            style = MaterialTheme.typography.labelSmall,
                            color = subtitleColor,
                            maxLines = 1,
                        )
                    }
                }
            }
            if (progressFraction != null) {
                Box(
                    modifier = Modifier
                        .align(Alignment.BottomStart)
                        .height(MetricStatProgressHeight)
                        .fillMaxWidth(progressFraction.coerceIn(0f, 1f))
                        .clip(MaterialTheme.shapes.extraSmall)
                        .background(progressFillColor),
                )
            }
        }
    }
}

File: lib/ui/components/metric_stat_card.dart
import 'package:flutter/material.dart';

/// A small dashboard stat tile (Distance, Total calories, Sleep, …). Faithful
/// port of the Kotlin `MetricStatCard` / the design-system `MetricStatCard`.
///
/// Layout: a 28dp accent-tinted icon circle (16dp glyph), then a title
/// ([labelMedium], on-surface-variant) over a value ([titleMedium], w600) with
/// an optional unit appended at medium weight, and an optional subtitle
/// ([labelSmall]). An optional 3dp accent progress underline is pinned to the
/// bottom edge and clipped to the rounded corners. Two tiles fit per row.
class MetricStatCard extends StatelessWidget {
  const MetricStatCard({
    super.key,
    required this.title,
    required this.value,
    required this.icon,
    required this.accentColor,
    this.unit,
    this.subtitle,
    this.message,
    this.showTitle = true,
    this.progress,
    this.onTap,
  });

  final String title;

  /// The formatted value; ignored when [message] is set.
  final String value;

  /// Optional unit appended after [value] at medium weight (e.g. `km`).
  final String? unit;
  final IconData icon;
  final Color accentColor;
  final String? subtitle;

  /// A no-data / loading message shown in place of the value, in the muted
  /// on-surface-variant colour.
  final String? message;
  final bool showTitle;

  /// Progress fraction in `[0, 1]`; when non-null, draws the accent underline.
  final double? progress;
  final VoidCallback? onTap;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final scheme = theme.colorScheme;
    final iconContainerColor = Color.alphaBlend(
      accentColor.withValues(alpha: 0.16),
      scheme.surfaceContainerHighest,
    );

    return Material(
      color: scheme.surfaceContainer,
      borderRadius: const BorderRadius.all(Radius.circular(12)),
      // Hard-edge clip keeps the progress underline inside the rounded corners
      // without the offscreen saveLayer cost of an antialiased clip.
      clipBehavior: Clip.hardEdge,
      child: InkWell(
        onTap: onTap,
        child: Stack(
          children: [
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 10),
              child: Row(
                children: [
                  Container(
                    width: 28,
                    height: 28,
                    decoration: BoxDecoration(
                      color: iconContainerColor,
                      shape: BoxShape.circle,
                    ),
                    child: Icon(icon, size: 16, color: accentColor),
                  ),
                  const SizedBox(width: 10),
                  Expanded(
                    child: Column(
                      mainAxisSize: MainAxisSize.min,
                      mainAxisAlignment: MainAxisAlignment.center,
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        if (showTitle)
                          Text(
                            title,
                            maxLines: 1,
                            overflow: TextOverflow.ellipsis,
                            style: theme.textTheme.labelMedium?.copyWith(
                              color: scheme.onSurfaceVariant,
                            ),
                          ),
                        _ValueLine(
                          value: value,
                          unit: unit,
                          message: message,
                          theme: theme,
                        ),
                        if (message == null &&
                            subtitle != null &&
                            subtitle!.trim().isNotEmpty)
                          Text(
                            subtitle!,
                            maxLines: 1,
                            overflow: TextOverflow.ellipsis,
                            style: theme.textTheme.labelSmall?.copyWith(
                              color: scheme.onSurfaceVariant,
                            ),
                          ),
                      ],
                    ),
                  ),
                ],
              ),
            ),
            if (progress != null)
              Positioned(
                left: 0,
                right: 0,
                bottom: 0,
                child: FractionallySizedBox(
                  alignment: Alignment.centerLeft,
                  widthFactor: progress!.clamp(0.0, 1.0),
                  child: Container(
                    height: 3,
                    color: accentColor.withValues(alpha: 0.55),
                  ),
                ),
              ),
          ],
        ),
      ),
    );
  }
}

class _ValueLine extends StatelessWidget {
  const _ValueLine({
    required this.value,
    required this.unit,
    required this.message,
    required this.theme,
  });

  final String value;
  final String? unit;
  final String? message;
  final ThemeData theme;

  @override
  Widget build(BuildContext context) {
    final scheme = theme.colorScheme;
    if (message != null) {
      return Text(
        message!,
        maxLines: 1,
        overflow: TextOverflow.ellipsis,
        style: theme.textTheme.titleMedium?.copyWith(
          fontWeight: FontWeight.w600,
          color: scheme.onSurfaceVariant,
        ),
      );
    }
    final base = theme.textTheme.titleMedium?.copyWith(
      fontWeight: FontWeight.w600,
      color: scheme.onSurface,
    );
    final hasUnit = unit != null && unit!.trim().isNotEmpty;
    return Text.rich(
      TextSpan(
        style: base,
        children: [
          TextSpan(text: value),
          if (hasUnit)
            TextSpan(
              text: ' $unit',
              style: base?.copyWith(fontWeight: FontWeight.w500),
            ),
        ],
      ),
      maxLines: 1,
      overflow: TextOverflow.ellipsis,
    );
  }
}
